File: backend/src/models/visita.js
'use strict';

const { Model, DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  class Visita extends Model {
    static associate(models) {
      Visita.belongsTo(models.Cliente, { foreignKey: 'cliente_id', as: 'cliente' });
      Visita.belongsTo(models.Imovel, { foreignKey: 'imovel_id', as: 'imovel' });
      Visita.belongsTo(models.User, { foreignKey: 'corretor_id', as: 'corretor' });
      Visita.belongsTo(models.User, { foreignKey: 'criado_por_id', as: 'criador' });
    }
  }

  Visita.init({
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
    cliente_id: { type: DataTypes.INTEGER, allowNull: false, references: { model: 'clientes', key: 'id' } },
    imovel_id: { type: DataTypes.INTEGER, allowNull: false, references: { model: 'imoveis', key: 'id' } },
    corretor_id: { type: DataTypes.INTEGER, allowNull: true, references: { model: 'users', key: 'id' } },
    criado_por_id: { type: DataTypes.INTEGER, allowNull: false, references: { model: 'users', key: 'id' } },
    tenant_id: { type: DataTypes.INTEGER, allowNull: true },
    data_visita: { type: DataTypes.DATE, allowNull: false },
    status: {
      type: DataTypes.ENUM('agendada', 'realizada', 'cancelada', 'reagendada'),
      defaultValue: 'agendada',
    },
    observacoes: { type: DataTypes.TEXT, allowNull: true },
    feedback_cliente: { type: DataTypes.TEXT, allowNull: true },
    nota_avaliacao: { type: DataTypes.INTEGER, allowNull: true, validate: { min: 1, max: 5 } },
  }, {
    sequelize,
    modelName: 'Visita',
    tableName: 'visitas',
    underscored: true,
    timestamps: true,
  });

  return Visita;
};

File: backend-go/internal/models/visita.go
package models

import "time"

// Visita espelha a tabela `visitas` — agendamento de visita de cliente a imóvel.
// Ver docs/migration/06-dashboards-vendas-config.md §"Visitas".
//
// GOTCHA (§9): o Node não checava dono em PUT/DELETE — no Go a autorização por
// tenant é garantida pelo escopo GORM (WithContext), mas o handler ainda deve
// aplicar a regra de negócio adicional quando exigida pelo produto.
type Visita struct {
	ID uint `gorm:"primaryKey" json:"id"`

	ClienteID    uint  `gorm:"column:cliente_id;index;not null" json:"cliente_id"`
	ImovelID     uint  `gorm:"column:imovel_id;index;not null" json:"imovel_id"`
	CorretorID   *uint `gorm:"column:corretor_id;index" json:"corretor_id,omitempty"`
	CriadoPorID  uint  `gorm:"column:criado_por_id;index" json:"criado_por_id"`
	TenantID     *uint `gorm:"column:tenant_id;index" json:"tenant_id,omitempty"`

	DataVisita time.Time `gorm:"column:data_visita" json:"data_visita"`
	Status     string    `gorm:"column:status;default:agendada" json:"status"`

	Observacoes     *string `gorm:"column:observacoes" json:"observacoes,omitempty"`
	FeedbackCliente *string `gorm:"column:feedback_cliente" json:"feedback_cliente,omitempty"`
	NotaAvaliacao   *int    `gorm:"column:nota_avaliacao" json:"nota_avaliacao,omitempty"` // 1-5

	CreatedAt time.Time `gorm:"column:created_at" json:"created_at"`
	UpdatedAt time.Time `gorm:"column:updated_at" json:"updated_at"`

	Cliente  *Cliente `gorm:"foreignKey:ClienteID;references:ID" json:"cliente,omitempty"`
	Imovel   *Imovel  `gorm:"foreignKey:ImovelID;references:ID" json:"imovel,omitempty"`
	Corretor *User    `gorm:"foreignKey:CorretorID;references:ID" json:"corretor,omitempty"`
}

func (Visita) TableName() string { return "visitas" }

// Status possíveis de uma visita.
const (
	VisitaStatusAgendada  = "agendada"
	VisitaStatusRealizada = "realizada"
	VisitaStatusCancelada = "cancelada"
	VisitaStatusReagendada = "reagendada"
)

var visitaStatusValidos = []string{
	VisitaStatusAgendada, VisitaStatusRealizada, VisitaStatusCancelada, VisitaStatusReagendada,
}

// IsVisitaStatusValido confere se o status pertence ao enum de visitas.
func IsVisitaStatusValido(status string) bool {
	for _, s := range visitaStatusValidos {
		if s == status {
			return true
		}
	}
	return false
}
